File: go_src/8.go
func myAtoi(str string) int {
	n := len(str)

	index, sign, result := 0, 1, 0
	for index < n && str[index] == ' ' {
		index++
	}

	if index == n {
		return 0
	}

	if str[index] == '+' {
		index++
	} else if str[index] == '-' {
		sign = -1
		index++
	}

	for ;index < n && str[index] >= '0' && str[index] <= '9'; index++ {
		digit := int(str[index] - '0')
		if result > (int(math.MaxInt32) - digit) / 10 {
			if sign == 1 {
				return int(math.MaxInt32)
			} else {
				return int(math.MinInt32)
			}
		}
		result = result * 10 + digit
	}

	return sign * result
}

File: java_src/8.java
class Solution {
   public int myAtoi(String str) {
        if (str == null || str.length() == 0) {
            return 0;
        }

        int index = 0, result = 0, len = str.length(), digit;
        boolean negative = true;
        char[] s = str.toCharArray();
        char c;
        while (index < len && s[index] == ' ') {
            index++;
        }
        if (index == len) {
            return 0;
        }
        if ((c = s[index]) == '+') {
            index++;
        } else if (c == '-') {
            negative = false;
            index++;
        }
        while (index < len && Character.isDigit(s[index])) {
            digit = s[index] - '0';
            if (result > (Integer.MAX_VALUE - digit) / 10) {
                return negative ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }
            result = result * 10 + digit;
            index++;
        }

        return negative ? result : -result;
    }
}
